blocks_reexecutor/blocks_reexecutor.go

```diff
@@ -199,3 +224,55 @@ func (s *BlocksReExecutor) Start(ctx context.Context, done chan struct{}) {
 func (s *BlocksReExecutor) StopAndWait() {
 	s.StopWaiter.StopAndWait()
 }
+
+func (s *BlocksReExecutor) commitStateAndVerify(statedb *state.StateDB, expected common.Hash, blockNumber uint64) (*state.StateDB, arbitrum.StateReleaseFunc, error) {
+	result, err := statedb.Commit(blockNumber, true)
+	if err != nil {
+		return nil, arbitrum.NoopStateRelease, err
+	}
+	if result != expected {
+		return nil, arbitrum.NoopStateRelease, fmt.Errorf("bad root hash expected: %v got: %v", expected, result)
+	}
+	sdb, err := state.New(result, statedb.Database(), nil)
+	if err == nil {
+		_ = s.db.TrieDB().Reference(result, common.Hash{})
```

Comment:
Ah makes sense! I've fixed it.

Reply:
nit:
alternatively, we can `Reference(result, common.Hash{})` before `state.New(result, ...)` => that should allow us to avoid locking on BlocksReExecutor level and save one mutex operation. There is already another rw mutex used in the hashdb implementation (pathdb doesn't have reference mechanism), so with the Reference first approach we shouldn't need extra locking.

Reference does nothing when called for trie node that is not cached in dirties cache => it doesn't hurt to call it before `state.New` that checks that the node exists anywhere (in dirties cache or cleans cache or in disk db).

Current approach should also work, so if we don't consider the extra mutex level too expensive then above is optional